### DISCUSSION TITLE
Datenaktualisierung tlw. unzuverlässig - woran kann es liegen?

### DISCUSSION THREAD

**Discussion by tttooommmeeekkk** [cyberthom42/vaillant-arotherm-plus#23] [Q&A] [answered]
Hallo,

vielleicht kann mir jemand helfen.
Woran liegt es, dass einige Daten aus dem ebusd nicht bzw. seeeehr unregelmäßg und damit unzuverlässig an den MQTT-Broker übergeben werden?

Auffällig ist es bei mir bspw. bei den Werten aus der hmu00
- _currentconsumedpower_
- _currentyieldpower_ .

Hier ein Beispiel:
![Bildschirmfoto vom 2024-11-05 17-37-31](https://github.com/user-attachments/assets/d78662a5-9159-4ff0-af5a-84c8af1c8bf3)

Im vorliegenden Fall ist der Wert für _currentyieldpower_ seit 16:00 bei 0,0kWh im ebusd; an den Broker wurde er bis jetzt (2h später 18:00 Uhr) noch nicht übergeben (gerade vor dem Absenden noch mal nachgeschaut).

Die Anzahl der zu übertragenden Werte habe ich schon drastisch reduziert (214 massages) und die Prio bei allen auf r1 gesetzt.
Andere Werte aus der HMU00 werden regelhafter an den mqtt-Broker übergeben und somit an meinen Client (Homeassistant) zur Auswertung übergeben.
Broker (Mosquitto) und Client (Homeassiatant) laufen jeweils in einem seperaten Docker auf meinem NAS. Die CPU- (unter 1%) und RAM-Auslastung (HA bei 1GB und MQTT bei 1MB) sind für beide Dienste komplett unauffällig. Und insgesamt läuft das NAS gemächlich bei 7% CPU und 31% RAM vor sich hin.

Was kann ich noch ändern, damit es zuverlässiger wird? Sind die knapp über 200 Werte noch zu viel für Prio1 (sollte ich noch bei einigen mit der Prio runter)?
Unschön, wenn in meinem Energy-Flow im Homeassistant noch ein Verbrauch für das Gerät angezeigt wird, obwohl die WP gerade nicht arbeitet.

**Comment by cyberthom42** [comment]
Also du kannst im ebusd log prüfen, ob die Daten überhaupt gesendet werden. Wobei ich mir nicht sicher bin, wie dein Setup aussieht. Wo läuft dein ebusd? Hast du einen Raspberry oder einen ESP? Die Push Intervalle bei MQTT sind für mich auch immer noch ein wenig Magie. Ich habe bisher nie die Abfrage-Intervalle gesehen die ich erwartet hätte. Ich frage die meißten Werte noch klassisch per API ab (über iobroker). Das funktioniert problemlos. Allerdings wäre der MQTT-Weg schon der richtigere. Ich bin auch bei der Umstellung, da ich die Werte sowohl in Telegraf (Influxdb) als auch iobroker und Homeassistant haben möchte. 
Du könntest dein Problem debuggen indem du dir ein kleines Abfrage-Script nur für diesen Wert bastelst, dann siehst du ob es mit dem MQTT zusammen hängt. Allerdings geht das auch nur wenn du den ebusd über den Raspberry laufen hast.

**Reply by chrizzzp** [reply, marked as answer]
@tttooommmeeekkk 

Ich habe auch mit den Poll-Priorititäten, die man über die CSVs definiert (r1, r2....), keine großen Erfolge gehabt. Ich löse es inzwischen so, daß ich per Home Assistant Automation die gewünschten Entitäten per MQTT abfrage, z.B. jede Minute. 

Man muß eine ‘MQTT publish action’ ausführen, z.B. mit dem MQTT topic

ebusd/hmu/currentconsumedpower/get

mit beliebigem Payload (z.B. 0) und QoS z.B. auf 2. Das lässt sich alles in HA per GUI (Automationen) einstellen.

Testen kann man die Aktion vorher in ‘Entwicklerwerkzeuge/Aktionen’ bevor man die Automation scharf schaltet. Ob es klappt sieht man im ebusd-Logfile oder in den Entitäten direkt.

Es gibt auch einige Entitäten, die z.B. minütlich vom VR921 abgefragt wird (falls man eins hat), die laufen dann auch über den Ebus und sollten die entsprechende Entität in HA aktualisieren. Ein Beispiel wäre:

`CurrentConsumedPowerWatt`

siehe hier: https://github.com/john30/ebusd-configuration/pull/330#issuecomment-2409046188

**Comment by tttooommmeeekkk** [comment]
Laut Logging (logareas: _all_) des ebusd (loglevel: _error_) ist nix auffällig.
Umstellung des Loggings auf loglevel _"info"_ hab ich mal testhalber kurz laufen lassen. Das hab ich aber gleich wieder umgestellt (auf error) - da kam (verständlicherweise) zu viel rein und das logfile wurde förmlich geflutet.

Hadrwaretechnisch ist mein ebusd ein _eBUS Adapter Shield v5_ (der vom ebus Adapter Team um john30) - also ein ESP32-C3 samt zusätzlich angestecktem Ethernet-Modul (das integrierte WLAN war/ist mir zu unsicher, was die Zuverlässigkeuit der Datenübertragung anbelangt).
Der ebusd läuft bei mir ebenfalls (recht ressourcenbindungs-sparsam) als Docker-Container auf meinem NAS (und beansprucht 0,05% CPU und 3MB RAM).


Hab jetzt mal HTTP in der ebusd-config aktiviert (bisher bin ich ohne HTTP unterwegs gewesen).
Auf dem HTTP_Port scheinen die Daten tatsächlich regelmäßiger anzukommen.

Ich habe jetzt mal den Vorschlag von @chrizzzp ausprobiert: Richtig toller Hinweis! Danke dafür!
Das funktioniert tatsächlich! Eine tolle Behelfslösung - ich habe es mal mit dem Dienstentwicklungstool (wie von dir vorgeschlagen) ausgetestet und werde dann wohl einige Sensoren in meine Automationen aufnehmen.
Wenn ich es richtig verstehe, wird damit ja auch kein Wert gesetzt (sondern lediglich der MQTT-Publish-Befehl mit "_get_" quasi zweckentfremdet), richtig?! Muss/sollte das "Retain"-Flag auf _"true"_ oder _"false"_ (ich bin da immer unsicher)?
Hast du bereits ein paar geeignette Kandidaten ausgemacht, die eines "get-Befehls" bedürfen oder holst holst du dir alle von dir benötigten Sensoren regelmäßig (welches Intervall?) ab?

@cyberthom42 : Die HTTP-Daten und damit ausgelesenen Sensoren kann man sicher auch irgendwie in HA impoertieren. Hast du da einen Hinweis, wie ich das bewerkstelligen kann?
Da muss ich wahrscheinlich die Sensoren mit `curl` in die configuration.yaml (bzw. in eine gesonderte ebus.yaml auslagen)
Gibt es da vielleicht ne Vorlage für Copy&Paste?
Wenn das möglich ist hätte ich eine weitere Möglichkeit (die vielleicht noch einfacher ist, weil keine zusätzlichen Atomationen wie bei der "MQTT-Publish-Get-Variante" anzulegen sind).
 
Dann könnte man mal schauen, was mir die Daten zuverlässig(er) aktualisiert.
Danke euch für eure Anregungen! Dann kann die anstehende Heizsaison hoffentlich endlich vernünftig mitgeloggt werden.

**Reply by chrizzzp** [reply]
> zweckentfremdet), richtig?! Muss/sollte das "Retain"-Flag auf _"true"_ oder _"false"_ (ich bin da immer unsicher)?

Habe ich bei mir auf "false". Man kriegt mit "Retain = true" halt die letzte bekannte MQTT-Message, wenn man ein Topic abonniert. Ist m.E. nicht allzu relevant für ebusd-Entitäten, da man immer aktuelle Werte haben will. So steht halt "unbekannt" als Wert, solange ebusd für die Entität noch nichts geliefert hat. Der einizige Vorteil wäre vielleicht, daß man beim Neustart von ebusd vielleicht noch die alten Werte (vor Neustart) drin hat (habe ich aber nicht getestet).

> Hast du bereits ein paar geeignette Kandidaten ausgemacht, die eines "get-Befehls" bedürfen oder holst holst du dir alle von dir benötigten Sensoren regelmäßig (welches Intervall?) ab?

Das ist sehr abhängig von dem was man mit ebusd machen will und welche Entitäten dafür kritisch sind. Ich greife z.B. per HA in die Steuerung von Heizkreisen ein, um z.B. PV-Überschuss gezielter nutzen zu können. 

Ich polle minütlich z.B. 

```
hmu CompressorModulation
hmu EnergyIntegral
hmu CurrentConsumedPower
hmu CurrentYieldPower
hmu Statuscode
ctlv2 SolarPump
ctlv2 Hc2Status
ctlcv2 Hc2ActualFlowtempdesired
ctlv2 Hc1Status
ctlv2 Hc1ActualFlowtempdesired
vwz00 ImmersionHeaterPower
vwz00 Statuscode
vwz00 Status00
ctlv2 Hc3Status
ctlv2 Hc3ExthwcOpmode
ctlv2 Hc3PumpStatus
ctlv2 Hc3ExtHwcActive
```

Einige Entitäten kommen ja auch automatisch (minütlich, teilweise alle 10 Sekunden) durch die Kommunikation der Komponenten untereinander und liefern z.B. Flussraten, Systemdruck, Vor- und Rücklauftemperaturen, Temperaturen der Sensoren, Energiewerte  etc. Diese braucht man dann nicht regelmässig pollen. 

Was den Import der Entitäten in HA angeht, ist hier aktuell was geschrieben worden (in einer weiter unten stehenden Antwort):
https://github.com/john30/ebusd/discussions/720#discussioncomment-11136874
Die automatische Aktualisierung wird damit m.E. aber nicht bewerkstelligt.

Zum Thema regelmäßige Aktualisierungen:
Man kann auch per [cron-Schedule](https://de.wikipedia.org/wiki/Cron) auf Linux-Ebene regelmässige Aktualisierungen der Entitäten über ebusctl erreichen, also z.B. jede Minute `ebusctl read -f -c hmu CurrentConsumedPower` ausführen lassen. Dadurch generiert man allerdings Abhängigkeiten vom Host-System, was bei einer Migration auf einen neuen Rechner vielleicht lästig ist.

**Reply by tttooommmeeekkk** [reply]
Das mit den `get` Befehlen per MQTT-Publish funktioniert echt gut, das reicht mir für meine Bedürfnisse. Danke noch nal für den Tipp.

Heizungsautomation nach PV-Ertrags-Rahmenbedingungen ist bei uns (Norddtl.) recht nutzlos.
Die paar Male (in der Übergangszeit), wo das Sinn macht, schalte ich noch manuell per HA (Heizen für gewisse Dauer, Warmwasser schnell). Von März bis Mitte Oktober läuft die Heizung hier eh nicht (nur Warmwasser - und da kommt i.d.R. immer genug Strom vom Dach) und die anderen Monate kommt einfach zu wenig vom Dach, als das es für Auto, Haus, Heizung & Akku reichen würde (da muss ich immer individuell abwägen).

Aber Automationen für die WP werden vielleicht auch für mich irgendwann (2025?) interessanter, wenn neben den flexiblen Stromtarifen auch die Netzentgelte vielleicht flexibel werden (iMSys vorausgesetzt). Aber das bleibt abzuwarten.
Da macht es dann auch Sinn, sich noch mal mit einer Aktualisierung der lokalen csv Dateien zu befassen (meine von jonesPD sind schon recht alt - da hat sich die letzte Zeit ja sicher Einiges getan und es gibt einige neue Sensoren, wie z.B. die von dir angesprochenen VR921). Ich sage mir: Aktuell hab ich alles was ichban Auswertungen & Funktionen brauche - never change a running system (wenn es nicht nötig ist).
